docs/man/man1/sim-torch.1
.TH SIM_TORCH "1" "2022-02-26" "LinuxCNC Documentation" "HAL Component"
.de TQ
.br
.ns
.TP \\$1
..

.SH NAME

sim-torch \- A simulated plasma torch
.SH SYNOPSIS
.HP
.B loadusr Wn sim-torch sim-torch
.SH DESCRIPTION

A simulated plasma torch for arc-ok testing.

.I VERSION:
.br
0.1


.SH PINS
.TP
.B sim-torch-rt.cut-noise-in\fR float in \fR(default: \fI0.75\fR)
the maximum amount of noise during cutting (volts)
.TP
.B sim-torch-rt.cycles-in\fR s32 in \fR(default: \fI200\fR)
the number of cycles that the arc voltage overshoots the cut voltage (cycles)
.TP
.B sim-torch-rt.on-delay-in\fR s32 in \fR(default: \fI10\fR)
the time from turn on until overshoot begins (cycles)
.TP
.B sim-torch-rt.offset-in\fR float in \fR
the cut voltage offset(volts)
.TP
.B sim-torch-rt.overshoot-in\fR s32 in \fR(default: \fI50\fR)
the percentage of the cut voltage that the arc voltage overshoots (percent)
.TP
.B sim-torch-rt.ramp-noise-in\fR float in \fR(default: \fI5\fR)
the maximum amount of noise during overshoot (volts)
.TP
.B sim-torch-rt.ramp-up-in\fR s32 in \fR(default: \fI80\fR)
percent of 'cycles_in' that the arc voltage ramps up (percent)
.TP
.B sim-torch-rt.start\fR bit in \fR
start the arc
.TP
.B sim-torch-rt.voltage-in\fR float in \fR(default: \fI100\fR)
the cut voltage (volts)
.TP
.B sim-torch-rt.voltage-out\fR float out \fR
output voltage (volts)

.SH AUTHOR
Phillip A Carter

.SH LICENSE
GPLv2 or greater


docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists

2005.12.20 - a new config picker has been written, if you start just the
             runscript this one helps you choose a config to run

2005.12.13 - new runscript, and config organization
             work towards the first release of EMC2

2005.xx.xx - EMC2 now supports tool-changing in complex scenarios

2005.xx.xx - classicladder now a HAL component, a software PLC

2004.xx.xx - HAL (the Hardware Abstraction Layer) is used for
	     addressing hardware drivers, and allows for simple configuration


#developers: this file should hold a history of improvements, bugfixes, etc
             it is useful for keeping track of development


src/hal/user_comps/sim-torch.py
'''
sim-torch.py

Copyright (C) 2022 Phillip A Carter

sim-torch.py is a simple plasma torch emulator

This program is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
'''

import hal
import math
import random
import time

# create sim-torch component
h = hal.component('sim-torch')
# create the hal pins
h.setprefix('sim-torch')
h.newpin('cut-noise-in', hal.HAL_FLOAT, hal.HAL_IN)
h.newpin('cycles-in', hal.HAL_S32, hal.HAL_IN)
h.newpin('on-delay-in', hal.HAL_S32, hal.HAL_IN)
h.newpin('offset-in', hal.HAL_FLOAT, hal.HAL_IN)
h.newpin('overshoot-in', hal.HAL_S32, hal.HAL_IN)
h.newpin('ramp-noise-in', hal.HAL_FLOAT, hal.HAL_IN)
h.newpin('ramp-up-in', hal.HAL_S32, hal.HAL_IN)
h.newpin('start', hal.HAL_BIT, hal.HAL_IN)
h.newpin('voltage-in', hal.HAL_FLOAT, hal.HAL_IN)
h.newpin('close', hal.HAL_BIT, hal.HAL_IN)
h.newpin('voltage-out', hal.HAL_FLOAT, hal.HAL_OUT)
h.newpin('void', hal.HAL_BIT, hal.HAL_IN)
h.newpin('void_ramp', hal.HAL_S32, hal.HAL_IN)
# setdefaults for input pins
h['cut-noise-in'] = 0.75
h['cycles-in'] = 200     #"the number of cycles that the arc voltage overshoots the cut voltage (cycles)"
h['on-delay-in']   = 10    #"the time from turn on until overshoot begins (cycles)"
h['overshoot-in']  = 50   #"the percentage of the cut voltage that the arc voltage overshoots (percent)"
h['ramp-noise-in'] = 5   #"the amount of noise during overshoot (volts)"
h['ramp-up-in']    = 80     #"percent of 'cycles-in' that the arc voltage ramps up (percent)"
h['voltage-in']    = 100    #"the cut voltage (volts)"
# component is ready
h.ready()

# working variables
angle = 0.0             # current ramp angle
current_cycle = 0       # current cycle
void_cycle = 0          # counter for void cycles
#void_ramp = 50          # voltage rise per cycle
dn_cycles = 0           # overshoot ramp down cycles
init = False            # initialized flag
initial_ramp = 10       # initial ramp up cycles
overshoot_max = 0.0     # maximum overshoot voltage
overshoot_start = 0.0   # overshoot start voltage
up_cycles = 0           # overshoot ramp up cycles

# return a random noise voltage
def random_noise(cycle, volume):
    if cycle % 5 == 0:
        random.seed()
        v = random.randrange(int(volume * -1000), int(volume * 1000))
        return v / 1000
    else:
        return 0

# main loop
try:
    while hal.component_exists('motmod') and not h['close']:
        # validate the inputs
        cut_noise = abs(h['cut-noise-in'])
        cycles = abs(h['cycles-in'])

        if h['on-delay-in'] < 0:
            on_delay = 0
        elif h['on-delay-in'] > 100:
            on_delay = 100
        else:
            on_delay = h['on-delay-in']

        if h['offset-in'] < -10:
            offset = -10
        elif h['offset-in'] > 10:
            offset = 10
        else:
            offset = h['offset-in']

        if h['overshoot-in'] < 0:
            overshoot = 0
        elif h['overshoot-in'] > 100:
            overshoot = 100
        else:
            overshoot = h['overshoot-in']

        ramp_noise = abs(h['ramp-noise-in'])

        if h['ramp-up-in'] < 0:
            ramp_up = 0
        elif h['ramp-up-in'] > 100:
            ramp_up = 100
        else:
            ramp_up = h['ramp-up-in']

        if h['voltage-in'] < 50:
            voltage = 50
        elif h['voltage-in'] > 150:
            voltage = 150
        else:
            voltage = h['voltage-in']

        # setup some variables
        up_cycles = cycles * (ramp_up * 0.01)
        dn_cycles = cycles - up_cycles
        overshoot_max = voltage * (overshoot * 0.01)
        overshoot_start = overshoot_max * 0.4

        if h['start']:
            # start the arc
            if(current_cycle < cycles):
                if current_cycle < on_delay:
                    # initial ramp up to cut voltage
                    h['voltage-out'] = (voltage + overshoot_start) / on_delay * current_cycle
                elif current_cycle <= up_cycles:
                    # ramp up to overshoot voltage
                    angle = math.atan2(overshoot_max - overshoot_start, up_cycles)
                    h['voltage-out'] = voltage + overshoot_start + current_cycle * math.tan(angle) + random_noise(current_cycle, ramp_noise)
                else:
                    # ramp down to cut voltage
                    angle = math.atan2(overshoot_max, dn_cycles)
                    h['voltage-out'] = voltage + overshoot_max - (current_cycle - up_cycles) * math.tan(angle) + random_noise(current_cycle, ramp_noise)
            else:
            # cut voltage voltage
                if h['void']:
                    void_cycle += 1
                    v = voltage + (h['void_ramp'] * void_cycle)
                    h['voltage-out'] = v if v < 300 else 300
                else:
                    h['voltage-out'] = voltage + offset + random_noise(current_cycle, cut_noise)
                    void_cycle = 0
            current_cycle += 1
        else:
            # stop the arc
            current_cycle = 0
            h['voltage-out'] = 0
        time.sleep(.001)
except KeyboardInterrupt:
    pass
except Exception as e:
    print(e)
